src/ddspdrum/torchmodule.py

```diff
@@ -486,12 +529,164 @@ def _forward(self, control_in: T, audio_in: T) -> T:
         assert (control_in >= 0).all() and (control_in <= 1).all()
 
         if (audio_in <= -1).any() or (audio_in >= 1).any():
-            normalize(audio_in)
+            util.normalize(audio_in)
 
-        audio_in = fix_length(audio_in, len(control_in))
+        audio_in = util.fix_length(audio_in, len(control_in))
         return control_in * audio_in
 
 
+class TorchNoise(TorchSynthModule):
+    """
+    Adds noise to a signal
+
+    Parameters
+    ----------
+    ratio (float): mix ratio between the incoming signal and the produced noise
+    **kwargs: see TorchSynthModule
+    """
+
+    def __init__(
+            self,
+            ratio: float = 0.25,
+            **kwargs
+    ):
+        super().__init__(**kwargs)
+        self.add_parameters(
+            [
+                TorchParameter(
+                    value=ratio,
+                    parameter_name="ratio",
+                    parameter_range=ParameterRange()
+                )
+            ]
+        )
+
+    def _forward(self, audio_in: T) -> T:
+        noise = self.noise_of_length(audio_in)
+        return util.crossfade(audio_in, noise, self.p("ratio"))
+
+    @staticmethod
+    def noise_of_length(audio_in: T) -> T:
+        return torch.rand_like(audio_in) * 2 - 1
+
+
+class TorchSynth(nn.Module):
+    """
+    Base class for synthesizers that combine one or more TorchSynthModules
+    to create a full synth architecture.
+
+    Parameters
+    ----------
+    sample_rate (int): sample rate to run this synth at
+    buffer_size (int): number of samples expected at output of child modules
+    """
+
+    def __init__(
+            self,
+            sample_rate: int = SAMPLE_RATE,
+            buffer_size: int = BUFFER_SIZE
+    ):
+        super().__init__()
+        self.sample_rate = T(sample_rate)
+        self.buffer_size = T(buffer_size)
+
+    def add_synth_modules(self, modules: Dict[str, TorchSynthModule]):
+        """
+        Add a set of named children TorchSynthModules to this synth. Registers them
+        with the torch nn.Module so that all parameters are recognized.
+
+        Parameters
+        ----------
+        modules (Dict): A dictionary of TorchSynthModule
+        """
+
+        for name in modules:
+            if not isinstance(modules[name], TorchSynthModule):
+                raise TypeError(f"{modules[name]} is not a TorchSynthModule")
+
+            if modules[name].sample_rate != self.sample_rate:
+                raise ValueError(f"{modules[name]} sample rate does not match")
```

Comment:
Should we be throwing an error here? Or perhaps just a warning and then set the sample rate of the child module being added to the sample rate of the Synth?

Reply:
I think error, because if the sample rates aren't matching there's surely some bigger problem earlier in the chain. My $0.02

Reply:
I think so, yeah (throw an error). Because if the rates aren't matching, then surely there's something off earlier in the code.

Reply:
Yeah I prefer complain loudly and fail

Reply:
In general, I like assertions (or valueerror raises) and think we should use them liberally